doc/mgmt/gnmi/master_arbitration.md

```diff
@@ -0,0 +1,331 @@
+- [Warmboot and Fastboot Design Impact](#warmboot-and-fastboot-design-impact)
+- [Restrictions / Limitations](#restrictions--limitations)
+- [Testing Requirements / Design](#testing-requirements--design)
+- [Open / Action items - if any](#open--action-items---if-any)
+
+### Revision
+
+| Rev  | Rev Date   | Author(s)          | Change Description |
+|------|------------|--------------------|--------------------|
+| v0.1 | 03/14/2023 | Tomek Madejski (Google), Sally Lyu (Google) | Initial version |
+
+### Scope
+
+This document describes the high level design of gNMI Master Arbitration. It provides a detailed explanation of the definition, as well as the implementation needed in the SONIC framework.
+
+
+### Definitions/Abbreviations
+
+gNMI - gRPC Network Management Interface (gNMI) is a protocol to manage and report a network element's configuration and operational data.
+
+gNMI client - a software application that is used to connect to a network element and manage its configuration and operational data. The gNMI client can be written in any language that supports gRPC.
+
+SDN - Software-defined networking (SDN) is a network architecture that uses software-based controllers or application programming interfaces (APIs) to communicate with underlying hardware infrastructure and direct traffic on a network.
+
+EID -  Election ID. A 128-bit integer. The election ID MUST be guaranteed to be monotonically increasing.But how it is done is out of scope of this document and implementation.
+
+Reference: https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-master-arbitration.md
+
+### Overview
+
+For high availability, a system may run multiple replicas of a gNMI client (controller). Among the replicas, only one client should be elected as master and do gNMI operations that mutate any state on the target. However, in the event of a network partition, there can be two or more replicas thinking themselves as master. For read-only RPCs, this is OK. But if they both call the `Set` RPC, the target may be incorrectly configured by the stale master. Therefore, "Master Arbitration" is needed when multiple clients exist.
+
+**Definition**
+
+When Master Arbitration feature is turned on, the gNMI server receives a proto with the election ID (EID) specified in it. The election ID is a 128-bit integer MUST be guaranteed to be monotonically increasing -- a new master's election ID MUST always be larger than its predecessor. The process through which election IDs are allocated to client replicas is out-of-scope of this specification.
+
+The client carries the election ID in all of its `Set` requests. When the target receives any `Set` request, it looks at the election ID carried with the requests and stores the largest ID it sees. An election ID that is equal to or greater than the currently stored one should be accepted as the master. On the other hand, Set requests with no election ID or with a smaller election ID are rejected and return an error.
```

Comment:
" An election ID that is equal to or greater than the currently stored one should be accepted as the master."
How does the arbitration work if there are two replicas that send the same (largest) election ID ?

Reply:
If two controllers are sending SET requests with the same EID, the switch will treat them as one controller and execute SET from both.

Reply:
Can there be more than 2 controllers?

Reply:
Also, how do we ensure conflicting requests are not received from multiple controllers with same EID? is there any use-case like this?

Reply:
> Can there be more than 2 controllers? if yes, please capture it accordingly in the diagrams.
Yes. There could be more. I can update the graph.

> Also, how do we ensure conflicting requests are not received from multiple controllers with same EID? is there any use-case like this?
If controllers send the same EID, basically this is the same senario as MA disabled case. However, controller manager should make sure that no EID should be unique.